File: src/cs/Ssh/SshServerSession.cs
// Copyright (c) Microsoft. All rights reserved.

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.DevTunnels.Ssh.Messages;
using Microsoft.DevTunnels.Ssh.Services;

namespace Microsoft.DevTunnels.Ssh;

/// <summary>
/// The server side of an SSH session. Extends the base `SshSession` class
/// to support host authentication.
/// </summary>
public class SshServerSession : SshSession
{
	private readonly ICollection<SshServerSession>? reconnectableSessions;

	/// <summary>
	/// Constructs a new server session.
	/// </summary>
	/// <param name="config">Session configuration.</param>
	/// <param name="trace">Trace source for the session.</param>
	public SshServerSession(
		SshSessionConfiguration config,
		TraceSource trace)
		: this(config, reconnectableSessions: null, trace)
	{
	}

	/// <summary>
	/// Constructs a new server session that is optionally capable of reconnecting.
	/// </summary>
	/// <param name="config">Session configuration.</param>
	/// <param name="reconnectableSessions">Collection that tracks server sessions
	/// available for reconnection.</param>
	/// <param name="trace">Trace source for the session.</param>
	public SshServerSession(
		SshSessionConfiguration config,
		ICollection<SshServerSession>? reconnectableSessions,
		TraceSource trace)
		: base(config, trace)
	{
		if (config == null) throw new ArgumentNullException(nameof(config));

		bool enableReconnect = config.ProtocolExtensions.Contains(
			SshProtocolExtensionNames.SessionReconnect);
		if (enableReconnect && reconnectableSessions == null)
		{
			throw new ArgumentException(
				"When reconnect is enabled, server sessions require a reference to a " +
					"shared collection to track reconnectable sessions.",
				nameof(reconnectableSessions));
		}
		else if (!enableReconnect && reconnectableSessions != null)
		{
			throw new ArgumentException(
				"When reconnect is not enabled, the reconnectable sessions collection " +
					"is not applicable.",
				nameof(reconnectableSessions));
		}

		this.reconnectableSessions = reconnectableSessions;
	}

	/// <summary>
	/// Gets or sets credentials and/or credential callbacks for authenticating the session.
	/// </summary>
	public SshServerCredentials Credentials { get; set; } = new SshServerCredentials();

	/// <summary>
	/// Event raised after the server has successfully authenticated the client.
	/// </summary>
	/// <remarks>
	/// This event may be used to trigger a server-initiated action after the client
	/// has authenticated.
	/// </remarks>
	public event EventHandler<EventArgs>? ClientAuthenticated;

	/// <summary>
	/// Event raised when the server session is reconnected.
	/// </summary>
	public event EventHandler<EventArgs>? Reconnected;

	/// <summary>
	/// Handles an incoming message. Can be overridden by subclasses to handle additional
	/// message types that are registered via <see cref="SshSessionConfiguration.Messages"/>.
	/// </summary>
	protected override Task HandleMessageAsync(
		SshMessage message, CancellationToken cancellation)
	{
		return message switch
		{
			ServiceRequestMessage m => HandleMessageAsync(m, cancellation),
			SessionRequestMessage m => HandleMessageAsync(m, cancellation),
			_ => base.HandleMessageAsync(message, cancellation),
		};
	}

	private async Task HandleMessageAsync(
		ServiceRequestMessage message, CancellationToken cancellation)
	{
		var service = ActivateService(message.ServiceName ?? string.Empty);
		if (service != null)
		{
			await SendMessageAsync(
				new ServiceAcceptMessage
				{
					ServiceName = message.ServiceName,
				},
				cancellation).ConfigureAwait(false);
		}
		else
		{
			throw new SshConnectionException(
				$"Service \"{message.ServiceName}\" not available.",
				SshDisconnectReason.ServiceNotAvailable);
		}
	}

	/// <summary>
	/// Handles server-specific session requests.
	/// </summary>
	internal override async Task HandleMessageAsync(
		SessionRequestMessage message, CancellationToken cancellation)
	{
		if (message.RequestType == ExtensionRequestTypes.SessionReconnect &&
			this.Config.ProtocolExtensions.Contains(SshProtocolExtensionNames.SessionReconnect))
		{
			var reconnectRequest = message.ConvertTo<SessionReconnectRequestMessage>();
			await this.ReconnectAsync(reconnectRequest, cancellation).ConfigureAwait(false);

			// ReconnectAsync() handles sending the response message.
			return;
		}

		await base.HandleMessageAsync(message, cancellation).ConfigureAwait(false);
	}

	/// <summary>
	/// Raises the ClientAuthenticated event.
	/// </summary>
	internal void HandleClientAuthenticated()
	{
		ClientAuthenticated?.Invoke(this, EventArgs.Empty);
	}

	/// <summary>
	/// Adds this session to the list of server sessions that are available
	/// for reconnection.
	/// </summary>
	internal override async Task EnableReconnectAsync(CancellationToken cancellation)
	{
		await base.EnableReconnectAsync(cancellation).ConfigureAwait(false);

		lock (this.reconnectableSessions!)
		{
			if (!this.reconnectableSessions.Contains(this))
			{
				this.reconnectableSessions.Add(this);
			}
		}
	}

	internal override bool OnDisconnected()
	{
		if (Reconnecting)
		{
			// Prevent closing the session while reconnecting.
			return true;
		}

		return base.OnDisconnected();
	}

	/// <summary>
	/// Attempts to reconnect the client to a disconnected server session.
	/// </summary>
	/// <remarks>
	/// If reconnection is successful, the current server session is disposed because the client
	/// gets reconnected to a different server session.
	/// </remarks>
	internal async Task ReconnectAsync(
		SessionReconnectRequestMessage reconnectRequest, CancellationToken cancellation)
	{
		if (this.reconnectableSessions == null)
		{
			throw new InvalidOperationException("Disconnected sessions collection " +
				"should have been initialized when reconnect is enabled.");
		}

		// Try to find the requested server session in the list of available disconnected
		// server sessions, by validating the reconnect token.
		SshServerSession? reconnectSession = null;
		lock (this.reconnectableSessions)
		{
			foreach (var reconnectableSession in this.reconnectableSessions)
			{
				if (reconnectableSession != this && VerifyReconnectToken(
					reconnectableSession.SessionId!,
					SessionId!,
					reconnectRequest.ClientReconnectToken))
				{
					reconnectSession = reconnectableSession;
					this.reconnectableSessions.Remove(reconnectSession);
					break;
				}
			}
		}

		if (reconnectSession == null || reconnectSession.IsClosed)
		{
			var message = "Requested reconnect session was not found " +
				"or the client's reconnect token was invalid.";
			Trace.TraceEvent(
				TraceEventType.Warning,
				SshTraceEventIds.ServerSessionReconnectFailed,
				$"{this} reconnect rejected: " + message);
			await SendMessageAsync(
				new SessionReconnectFailureMessage
				{
					ReasonCode = SshReconnectFailureReason.SessionNotFound,
					Description = message,
				},
				cancellation).ConfigureAwait(false);
			return;
		}

		var messagesToResend = reconnectSession.Protocol!.GetSentMessages(
			reconnectRequest.LastReceivedSequenceNumber + 1);
		if (messagesToResend == null)
		{
			// Messages are not available from requested sequence number.
			// Restore the current session protocol and put the old session back in the collection.
			lock (this.reconnectableSessions)
			{
				this.reconnectableSessions.Add(reconnectSession);
			}

			var message = "Server is unable to re-send messages requested by the client.";
			Trace.TraceEvent(
				TraceEventType.Warning,
				SshTraceEventIds.ServerSessionReconnectFailed,
				$"{this} reconnect rejected: " + message);
			await SendMessageAsync(
				new SessionReconnectFailureMessage
				{
					ReasonCode = SshReconnectFailureReason.ServerDroppedMessages,
					Description = message,
				},
				cancellation).ConfigureAwait(false);
			return;
		}

		var reconnectToken = CreateReconnectToken(reconnectSession.SessionId!, SessionId!);
		await SendMessageAsync(
			new SessionReconnectResponseMessage
			{
				ServerReconnectToken = reconnectToken,
				LastReceivedSequenceNumber = reconnectSession.Protocol!.LastIncomingSequence,
			},
			cancellation).ConfigureAwait(false);

		try
		{
			reconnectSession.Reconnecting = true;

			// Ensure the old connection is disconnected before switching over to the new one.
			reconnectSession.Protocol.Dispose();
			while (reconnectSession.IsConnected)
			{
				await Task.Delay(5, cancellation).ConfigureAwait(false);
			}

			// Move this session's protocol instance over to the reconnected session.
			reconnectSession.Protocol = Protocol;
			reconnectSession.Protocol!.KeyExchangeService =
				reconnectSession.GetService<KeyExchangeService>();
			reconnectSession.Protocol!.TraceChannelData = reconnectSession.Config.TraceChannelData;
			Protocol = null;

			// Re-send the lost messages that the client requested.
			foreach (var message in messagesToResend)
			{
				await reconnectSession.Protocol.SendMessageAsync(message, cancellation)
					.ConfigureAwait(false);
			}

			// Now this server session is invalid because the client reconnected to another one.
			Dispose(new SshConnectionException("Reconnected.", SshDisconnectReason.None));
		}
		finally
		{
			reconnectSession.Reconnecting = false;
		}

		lock (this.reconnectableSessions)
		{
			this.reconnectableSessions.Add(reconnectSession);
		}

		reconnectSession.Metrics.AddReconnection();

		// Restart the message loop for the reconnected session.
		reconnectSession.ProcessMessages();

		Trace.TraceEvent(
			TraceEventType.Information,
			SshTraceEventIds.ServerSessionReconnecting,
			$"{this} reconnected {reconnectSession}. Re-sent {messagesToResend.Count} dropped messages.");

		// Notify event listeners about the successful reconnection.
		reconnectSession.Reconnected?.Invoke(reconnectSession, EventArgs.Empty);
	}

	protected override void Dispose(bool disposing)
	{
		if (disposing)
		{
			if (this.reconnectableSessions != null)
			{
				lock (this.reconnectableSessions)
				{
					this.reconnectableSessions.Remove(this);
				}
			}
		}

		base.Dispose(disposing);
	}
}

File: src/ts/ssh/sshServerSession.ts
//
//  Copyright (c) Microsoft Corporation. All rights reserved.
//

import { SshSession, ExtensionRequestTypes } from './sshSession';
import { CancellationToken, Emitter } from 'vscode-jsonrpc';
import {
	SshDisconnectReason,
	ServiceRequestMessage,
	ServiceAcceptMessage,
	SessionReconnectRequestMessage,
	SessionRequestFailureMessage,
	SessionReconnectResponseMessage,
	SessionRequestMessage,
	SessionReconnectFailureMessage,
	SshReconnectFailureReason,
} from './messages/transportMessages';
import { SshConnectionError } from './errors';
import { SshSessionConfiguration, SshProtocolExtensionNames } from './sshSessionConfiguration';
import { SshServerCredentials } from './sshCredentials';
import { TraceLevel, SshTraceEventIds } from './trace';

/**
 * The server side of an SSH session. Extends the base `SshSession` class
 * to support host authentication.
 */
export class SshServerSession extends SshSession {
	private readonly reconnectableSessions?: SshServerSession[];

	public constructor(config: SshSessionConfiguration, reconnectableSessions?: SshServerSession[]) {
		super(config, false);

		const enableReconnect = config.protocolExtensions.includes(
			SshProtocolExtensionNames.sessionReconnect,
		);
		if (enableReconnect && !reconnectableSessions) {
			throw new Error(
				'When reconnect is enabled, server sessions require a reference to a ' +
					'shared collection to track reconnectable sessions.',
			);
		} else if (!enableReconnect && reconnectableSessions) {
			throw new Error(
				'When reconnect is not enabled, the reconnectable sessions collection ' +
					'is not applicable.',
			);
		}

		this.reconnectableSessions = reconnectableSessions;
	}

	private readonly clientAuthenticatedEmitter = new Emitter<void>();
	public readonly onClientAuthenticated = this.clientAuthenticatedEmitter.event;

	private readonly reconnectedEmitter = new Emitter<void>();
	public readonly onReconnected = this.reconnectedEmitter.event;

	/**
	 * Gets or sets credentials and/or credential callbacks for authenticating the session.
	 */
	public credentials: SshServerCredentials = { publicKeys: [] };

	/* @internal */
	protected async handleServiceRequestMessage(
		message: ServiceRequestMessage,
		cancellation?: CancellationToken,
	): Promise<void> {
		const service = this.activateService(message.serviceName!);
		if (service) {
			const acceptMessage = new ServiceAcceptMessage();
			acceptMessage.serviceName = message.serviceName;
			await this.sendMessage(acceptMessage, cancellation);
		} else {
			throw new SshConnectionError(
				`Service "${message.serviceName}" not available.`,
				SshDisconnectReason.serviceNotAvailable,
			);
		}
	}

	/* @internal */
	protected async handleRequestMessage(
		message: SessionRequestMessage,
		cancellation?: CancellationToken,
	): Promise<void> {
		if (
			message.requestType === ExtensionRequestTypes.sessionReconnect &&
			this.config.protocolExtensions?.includes(SshProtocolExtensionNames.sessionReconnect)
		) {
			const reconnectRequest = message.convertTo(new SessionReconnectRequestMessage());
			await this.reconnect(reconnectRequest, cancellation);

			// reconnect() handles sending the response message.
			return;
		}

		await super.handleRequestMessage(message, cancellation);
	}

	/* @internal */
	public handleClientAuthenticated() {
		this.clientAuthenticatedEmitter.fire();
	}

	/* @internal */
	public async enableReconnect(cancellation?: CancellationToken) {
		await super.enableReconnect(cancellation);

		if (!this.reconnectableSessions!.includes(this)) {
			this.reconnectableSessions!.push(this);
		}
	}

	/* @internal */
	public handleDisconnected(): boolean {
		if (this.reconnecting) {
			// Prevent closing the session while reconnecting.
			return true;
		}

		return super.handleDisconnected();
	}

	/**
	 * Attempts to reconnect the client to a disconnected server session.
	 *
	 * If reconnection is successful, the current server session is disposed because the client
	 * gets reconnected to a different server session.
	 */
	/* @internal */
	public async reconnect(
		reconnectRequest: SessionReconnectRequestMessage,
		cancellation?: CancellationToken,
	) {
		if (!this.reconnectableSessions) {
			throw new Error(
				'Disconnected sessions collection ' +
					'should have been initialized when reconnect is enabled.',
			);
		}

		// Try to find the requested server session in the list of available disconnected
		// server sessions, by validating the reconnect token.
		let reconnectSession: SshServerSession | undefined;
		for (const reconnectableSession of this.reconnectableSessions) {
			if (
				reconnectableSession !== this &&
				(await this.verifyReconnectToken(
					reconnectableSession.sessionId!,
					this.sessionId!,
					reconnectRequest.clientReconnectToken ?? Buffer.alloc(0),
				))
			) {
				reconnectSession = reconnectableSession;
				this.reconnectableSessions.splice(
					this.reconnectableSessions.indexOf(reconnectSession),
					1,
				);
				break;
			}
		}

		if (!reconnectSession || reconnectSession.isClosed) {
			const message =
				'Requested reconnect session was not found or ' +
				'the reconnect token provided by the client was invalid.';
			this.trace(
				TraceLevel.Error,
				SshTraceEventIds.serverSessionReconnectFailed,
				`Reconnect failed: ${message}`,
			);
			const failure = new SessionReconnectFailureMessage();
			failure.reasonCode = SshReconnectFailureReason.sessionNotFound;
			failure.description = message;
			await this.sendMessage(failure, cancellation);
			return;
		}

		const messagesToResend = reconnectSession.protocol!.getSentMessages(
			(reconnectRequest.lastReceivedSequenceNumber ?? 0) + 1,
		);
		if (!messagesToResend) {
			// Messages are not available from requested sequence number.
			// Restore the current session protocol and put the old session back in the collection.
			this.reconnectableSessions.push(reconnectSession);

			const message = 'Server is unable to re-send messages requested by the client.';
			this.trace(
				TraceLevel.Error,
				SshTraceEventIds.serverSessionReconnectFailed,
				`Reconnect failed: ${message}`,
			);
			const failure = new SessionReconnectFailureMessage();
			failure.reasonCode = SshReconnectFailureReason.serverDroppedMessages;
			failure.description = message;
			await this.sendMessage(failure, cancellation);
			return;
		}

		const responseMessage = new SessionReconnectResponseMessage();
		responseMessage.serverReconnectToken = await this.createReconnectToken(
			reconnectSession.sessionId!,
			this.sessionId!,
		);
		responseMessage.lastReceivedSequenceNumber = reconnectSession.protocol!.lastIncomingSequence;
		await this.sendMessage(responseMessage, cancellation);

		try {
			reconnectSession.reconnecting = true;

			// Ensure the old connection is disconnected before switching over to the new one.
			reconnectSession.protocol?.dispose();
			while (reconnectSession.isConnected) {
				await new Promise<void>((resolve) => setTimeout(() => resolve(), 5));
			}

			// Move this session's protocol instance over to the reconnected session.
			reconnectSession.protocol = this.protocol;
			reconnectSession.protocol!.kexService = reconnectSession.kexService;
			this.protocol = undefined;

			// Re-send the lost messages that the client requested.
			for (const message of messagesToResend) {
				await reconnectSession.sendMessage(message, cancellation);
			}

			// Now this server session is invalid because the client reconnected to another one.
			this.dispose(new SshConnectionError('Reconnected.', SshDisconnectReason.none));
		} finally {
			reconnectSession.reconnecting = false;
		}

		this.reconnectableSessions.push(reconnectSession);

		reconnectSession.metrics.addReconnection();

		// Restart the message loop for the reconnected session.
		reconnectSession.processMessages().catch((e) => {
			this.trace(
				TraceLevel.Error,
				SshTraceEventIds.unknownError,
				`Unhandled error processing messages: ${e.message}`,
				e,
			);
		});

		this.trace(
			TraceLevel.Info,
			SshTraceEventIds.serverSessionReconnecting,
			`${reconnectSession} reconnected. Re-sent ${messagesToResend.length} dropped messages.`,
		);

		// Notify event listeners about the successful reconnection.
		reconnectSession.reconnectedEmitter.fire();
	}

	public dispose(): void;

	/* @internal */
	public dispose(error?: Error): void;

	public dispose(error?: Error): void {
		if (this.reconnectableSessions) {
			const index = this.reconnectableSessions.indexOf(this);
			if (index >= 0) {
				this.reconnectableSessions.splice(index, 1);
			}
		}

		super.dispose(error);
	}
}
